adjusted consoling, added tests, added bash.sh

Ali-sbp opened the pull request:
minor improvement on test consoling. 

added test_stats for baseline easy and hard for improved testability. 

added bash.sh for quicker project making and remaking. 

~~all around quality of life improvements based on my personal experience 😃

Ali-sbp committed:
Asd

Ali-sbp committed:
clean up, added new tests, adjusted cmake, adjusted consoling for tests

Ali-sbp committed:
added human players to the core, added test my_player vs human which helps debugging

Ali-sbp commented:
new commit : 

Added human player to the game core, now every bot can be played with vs human players. 

added a test to test my_player vs human which incredibly helps debugging the AI logic

Ali-sbp committed:
quick fix on win length

Ali-sbp committed:
changed display_board function to exactly match the existing testfiles

vsennov commented on bash.sh:
Cannot see the purpose of this file. How does running this differ from:

```sh
cd build # done once
cmake -DBUILD_TTTCORE=PREBUILT .. # only when new sources are added
make # when building
```

I do not think this file should be included in main repo.

Ali-sbp replied:
purpose was just simplicity, I have made this project probably somewhere around 1000 times and it just cuts into the typing time basically. that was the whole purpose to show (remind others maybe that this could be done) 

but if this is something you are against, I will remove it.

vsennov replied:
It is common to keep a set of such personal utils when working with code, but I do not think this particular one should be share and merged in main code. Please, remove it.

vsennov commented on libs/libtttcore.a:
We are not uploading prebuilt libs in source code, please, remove it.

Ali-sbp replied:
understood , will do.

vsennov commented on src/player/human_player.cpp:
```diff
@@ -0,0 +1,95 @@
+    return result;
+}
+
+void HumanPlayer::display_board(const State& state) const {
```
Seems like this code is also used in tests, so I think it deserves to be a global function, that will be used in tests and also in this class.

Ali-sbp replied:
understood, so adding the global function to observer also in place of its current console writing function, correct?

vsennov replied:
Yes.

vsennov commented on tests/CMakeLists.txt:
```diff
@@ -4,10 +4,28 @@ add_executable(test_my_player test_my_player.cpp)
+#if((BUILD_TTTCORE STREQUAL "FULL") OR (BUILD_TTTCORE STREQUAL "PREBUILT"))
+add_executable(test_my_player_vs_human test_my_player_vs_human.cpp)
+target_link_libraries(test_my_player_vs_human tttplayer)
+add_test(NAME test_my_player_vs_human COMMAND ./test_my_player_vs_human)
```
This test requires shell interaction, doesn't it? `add_test` adds this executable to automatic tests, that run without interaction, I don't get, how is it going to run :(

Ali-sbp replied:
I am not certain but I ran it just like this and It seemed to run fine? 
I literally tested it and it works with simple my_player player.

vsennov replied:
Did you run `make test`?

Ali-sbp replied:
no , I just build the whole project, I cloned this repo and ran the bash.sh and it worked

vsennov replied:
`make test` runs all tests, that have been added via `add_test` in cmake, and then prints summary. I'm not sure how it will do if program expect stdin, probably, it immediately sends EOF to the program, that will crash your test.

Ali-sbp replied:
I will add all of the requested changes and will get back to you on this one. can't say in the moment.

vsennov commented on CMakeLists.txt:
```diff
@@ -37,7 +37,7 @@ endif()
 
 # NOTE: add source files for your players here
-set(player_src src/player/my_player.cpp src/player/my_observer.cpp)
+set(player_src src/player/human_player.cpp src/player/my_player.cpp src/player/my_observer.cpp)
```
I think, your human player is more related to testing and manual debugging, maybe it should be placed in `tests/`? `tttplayer` represents your final solution, human player is more like auxiliary part to this.

Ali-sbp replied:
if you prefer it that way then sure, I can move it to tests along with the tests.

vsennov commented on src/player/human_player.hpp:
If this file didn't mentioned in README, it is nice to add a little doc comment in the begging with short description and example usage.

Ali-sbp replied:
noted, will add them on next commit.

vsennov commented on tests/test_stats_baseline_easy.cpp:
This code repeats 3 times in different stats tests. I think, we can make common header, with all the time counting stuff and one static function `int test_stats(int n_times, IPlayer *p1, IPlayer *p2)`. Then in 3 different test you just create players and run this function.

Ali-sbp replied:
great idea, it is a lot cleaner, will add this.

vsennov commented on tests/CMakeLists.txt:
```diff
@@ -4,10 +4,28 @@ add_executable(test_my_player test_my_player.cpp)
 target_link_libraries(test_my_player tttplayer)
 add_test(NAME smoke_test_player COMMAND ./test_my_player)
 
+if((BUILD_TTTCORE STREQUAL "FULL") OR (BUILD_TTTCORE STREQUAL "PREBUILT"))
```
I initially thought, that all tests that require baseline player will be included in one big `if` body, I think this is the more elegant way to do it.

Ali-sbp replied:
if I understood you correctly, I will add all of them in a one big if clause, works that way and it is more elegant I totally agree.

vsennov reviewed (changes requested):
All in all, I think we should do the following changes:

1. Move your beautiful board printing routine to some global function, then use it in human player and tests. You can place it in `my_observer.hpp` maybe?
2. Separate all tests that require manual interaction in something like `tests/manual/`. I think `human_player` can be also define in this section and linked to the tests requiring it. Do not this tests in autotests using `add_test`.
3. Move common stats testing routine as static function in common header and reuse it in all stats tests.
4. Add a little doc comment or mention your manual testing utilities in readme.
5. Remove binary libs and `bash.sh`

Ali-sbp committed:
test

Ali-sbp committed:
test_stats.hpp utils.cpp

added test_stats.hpp which includes the main test file, added utils.cpp and utils.hpp but i think we need to add them to libtttcore.a , tried to add the function to the game.gpp but still couldn't call it, since I think i cant add them to the lib

vsennov commented on src/player/my_observer.cpp:
```diff
@@ -41,6 +42,9 @@ void ConsoleWriter::handle_event(const State &, const Event &event) {
               << event.data.move.x << ", " << event.data.move.y << ")"
               << std::endl;
+    //using printing function here:
+    //cant call the function from observer cause need to be built into libtttcore.a ?
```
Well, my proposal was to declare this function in `my_observer.hpp` and define in `my_observer.cpp`, so this function can be linked within player lib.

vsennov commented on .vscode/settings.json:
I think, this should be put in `.gitignore`.

vsennov commented on src/core/game.hpp:
```diff
@@ -5,6 +5,8 @@
 
 namespace ttt::game {
 
+  void print_game_state(const State& state);
```
Why do you need to define this function twice?

I suppose, because you cannot recompile core library, that's why I told you to put this function in `my_observer.hpp`.

vsennov commented on src/player/human_player.cpp:
```diff
@@ -0,0 +1,95 @@
+    return result;
+}
+//asd
+void HumanPlayer::display_board(const State& state) const {
```
Why didn't you use your global function here?

vsennov reviewed (changes requested):
So, my initial thought was to place this printing function in `player/my_observer.hpp/cpp` as global function, so that this change could be done without recompiling core lib and all the students feel free customizing it — they would not be able to if changing this function requires recompiling our core lib.

I still think this is better place for this function to live.

So, I would ask you to:

1. Move declaration of this printing function in  `player/my_observer.hpp` (as a global function) and definition in `player/my_observer.cpp`.
2. Use this function in `HumanPlayer` and in tests.

There are also some files staged, that we agreed to delete from git (libs) and some changes about not adding manual tests in ctest.

Commit:
added all the required changes

Ali-sbp commented:
implemented all the requested changes on the new commit : 
1.removed requested files. 
2.moved human player to test
3.added the global print function to observer and used it everywhere possible. 
4.added the global test file and used it everywhere
5.removed test_vs_human from make tests to stop causing it from auto running it while using make test (because it needs to be played)

Commit:
Remove .vscode from repo

Commit:
Remove .DS_Store from repo

Ali-sbp commented:
also added contents to README.md , search for Игрок-человек для тестирования
